### PR TITLE
Implement gRPC server

### LINKED ISSUE #2: Implement gRPC server

### PR DESCRIPTION
**Pull request by asutula**
So far I have `AddFile` and `GetFile` implemented and working over gRPC with integration tests.

For now, I want some feedback on the way the code is written, Go tips, etc. I know this is pretty rough at this point, but I'd like to get feedback before writing the rest of it.

Closes #2

### CODE REVIEW

File: ipfs-lite/ipfs_lite.proto [hunk 1]
```diff
@@ -180,7 +178,7 @@ message HashOnReadResponse {
 }
 
 service IpfsLite {
-    rpc AddFile(stream AddFileRequest) returns (AddFileResponse) {}
+    rpc AddFile(AddFileRequest) returns (AddFileResponse) {}
```
Comment (asutula) [inline ipfs-lite/ipfs_lite.proto]: Changed this to not-a-stream for now just to simplify, will add back in later.

File: server/server.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+	grpcServer := grpc.NewServer()
+	server := &ipfsLiteServer{
+		peer: peer,
+	}
+	pb.RegisterIpfsLiteServer(grpcServer, server)
+	err = grpcServer.Serve(lis)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (jsign) [inline server/server.go]: A possible simplification:
`return grpcServer.Serve(lis)`

File: server/server.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	pb "textile.io/grpc-ipfs-lite/ipfs-lite"
+)
+
+type ipfsLiteServer struct {
+	pb.UnimplementedIpfsLiteServer
+
+	// should this be a file var or here?
+	peer *ipfslite.Peer
```
Comment (jsign) [inline server/server.go]: By `file var` you mean a package-scope variable?
Considering that is received in `StartServer` seems to be more bundled with this struct. Theoretically may make sense to call `StarServer` multiple times (in real-word, or in tests) and a file var may be handcuffing.
TL;DR: seems good for me as it is

File: server/server.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	respBlock := pb.Block{
+		RawData: node.RawData(),
+		Cid:     node.Cid().String(),
+	}
+
+	respLinks := []*pb.Link{}
```
Comment (jsign) [inline server/server.go]: Possible improvement:
```suggestion
 	respLinks :=make([]*pb.Link, len(node.Links()))
```
Which allocates the right size of the slice in advance instead of making grow on-demand while `append`ing to it. (slices grow twice the size everytime they get full). Making this change would mean also changing the `append` to be index-assigned: `respLinks[i] = &respLink` where `i` should be de ignored `range` var.

Another option is defining the slice capacity but with zero-length. That way you can use `append` and won't grow since it can stretch the length without reaching max capacity.

File: server/server.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	reader, err := s.peer.GetFile(ctx, cid)
+	if err != nil {
+		return nil, err
+	}
+
+	buffer, err := ioutil.ReadAll(reader)
```
Comment (jsign) [inline server/server.go]: The slack conversation! :smile: 
Maybe adding a `// ToDo` here since this is temporal... I guess has to do with your comment on making the api not streamed based to move forward! :+1:

File: server/server_test.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/ipfs/go-log"
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	"github.com/multiformats/go-multiaddr"
+	"google.golang.org/grpc"
+	pb "textile.io/grpc-ipfs-lite/ipfs-lite"
+)
+
+var (
+	client      pb.IpfsLiteClient
+	stringToAdd string = "hola"
```
Comment (jsign) [inline server/server_test.go]: :es:  :100: :100:

**Review by jsign — APPROVED** [review]
LGTM
Since you want some Go advice, I mentioned minor things that doesn't affect code correctness :+1:

File: server/server.go (hunk 3 above)
Comment (asutula) [reply server/server.go]: Cool, thanks!

File: server/server_test.go (hunk 6 above)
Comment (asutula) [reply server/server_test.go]: :)

File: server/server.go (hunk 4 above)
Comment (asutula) [reply server/server.go]: Nice, making that change for sure

File: server/server.go [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+	pb.RegisterIpfsLiteServer(grpcServer, server)
+	return grpcServer.Serve(lis)
+}
+
+type addFileResult struct {
+	Node format.Node
+	Err  error
+}
+
+func addFile(ctx context.Context, peer *ipfslite.Peer, addParams *pb.AddParams, reader io.Reader, ch chan addFileResult) {
```
Comment (asutula) [inline server/server.go:45]: This method gets forked off on a goroutine and sends the result back on the cannel that is passed it

File: server/server.go [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+	params := &ipfslite.AddParams{
+		Layout:    addParams.GetLayout(),
+		Chunker:   addParams.GetChunker(),
+		RawLeaves: addParams.GetRawLeaves(),
+		Hidden:    addParams.GetHidden(),
+		Shard:     addParams.GetShared(),
+		NoCopy:    addParams.GetNoCopy(),
+		HashFun:   addParams.GetHashFun(),
+	}
+	node, err := peer.AddFile(ctx, reader, params)
```
Comment (asutula) [inline server/server.go:56]: It will block here until the `reader` completes, could take a while. The reader is being "driven" below by the `io.PipeWriter`.

File: server/server.go [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+	if err != nil {
+		ch <- addFileResult{Err: err}
+		return
+	}
+	ch <- addFileResult{Node: node}
+}
+
+func (s *ipfsLiteServer) AddFile(srv pb.IpfsLite_AddFileServer) error {
+	req, err := srv.Recv()
```
Comment (asutula) [inline server/server.go:65]: Just receive the first stream event from the client

File: server/server.go [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+	if err != nil {
+		return err
+	}
+	var addParams *pb.AddParams
+	switch payload := req.GetPayload().(type) {
+	case *pb.AddFileRequest_AddParams:
```
Comment (asutula) [inline server/server.go:71]: The first stream event must contain the `AddParams` or else we error out.

File: server/server.go [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+		addParams = payload.AddParams
+	default:
+		return fmt.Errorf("expexted AddParams for AddFileRequest.Payload but got %T", payload)
+	}
+
+	reader, writer := io.Pipe()
+
+	addFileChannel := make(chan addFileResult)
+	go addFile(srv.Context(), s.peer, addParams, reader, addFileChannel)
```
Comment (asutula) [inline server/server.go:80]: Create the channel we'll receive the results of creating the `Node` on and fork off that call in a goroutine

File: server/server.go [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+
+	for {
+		req, err := srv.Recv()
```
Comment (asutula) [inline server/server.go:83]: Now we listen for all subsequent stream events from the client. They should be file chunks.

File: server/server.go [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+		if err == io.EOF {
+			writer.Close()
+			break
```
Comment (asutula) [inline server/server.go:86]: If we get `EOF` from the client, close the writer so the ongoing call to `AddFile` will finish

File: server/server.go [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+		} else if err != nil {
+			writer.CloseWithError(err)
+			break
```
Comment (asutula) [inline server/server.go:89]: If we get any other error from the client, close the writer with that error. I'm thinking that error will be returned from `AddFile` and subsequently be received over the channel and returned in this method.

File: server/server.go [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+		}
+		switch payload := req.GetPayload().(type) {
+		case *pb.AddFileRequest_Chunk:
```
Comment (asutula) [inline server/server.go:92]: We're only interested in file chunks, anything else is an error.

File: server/server.go [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+			_, writeErr := writer.Write(payload.Chunk)
+			if writeErr != nil {
+				return writeErr
+			}
+		default:
+			return fmt.Errorf("expected Chunk for AddFileRequest.Payload but got %T", payload)
+		}
+	}
+
+	addFileResult := <-addFileChannel
```
Comment (asutula) [inline server/server.go:102]: Once the client is done sending data, get the result of the `AddFile` goroutine from the channel.

File: server/server.go [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+	if addFileResult.Err != nil {
+		return addFileResult.Err
+	}
+
+	respNode, err := nodeToPbNode(addFileResult.Node)
+	if err != nil {
+		return err
+	}
+
+	return srv.SendAndClose(&pb.AddFileResponse{Node: respNode})
```
Comment (asutula) [inline server/server.go:112]: Send the resulting node back to the client.

File: server/server.go [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+	}
+	node, err := peer.AddFile(ctx, reader, params)
+	if err != nil {
+		ch <- addFileResult{Err: err}
+		return
+	}
+	ch <- addFileResult{Node: node}
+}
+
+func (s *ipfsLiteServer) AddFile(srv pb.IpfsLite_AddFileServer) error {
```
Comment (asutula) [inline server/server.go:64]: This is the actual server impl method for adding a file

File: server/server.go (hunk 15 above)
Comment (asutula) [reply server/server.go:92]: Write the chunk to our io pipe writer on the line below.

File: ipfs-lite/ipfs_lite.proto [hunk 19]
```diff
@@ -182,37 +214,29 @@ message HashOnReadResponse {
...
-    rpc ResolveLink(ResolveLinkRequest) returns (ResolveLinkResponse) {}
+    rpc HasBlock(HasBlockRequest) returns (HasBlockResponse) {}
 
-    rpc Resolve(ResolveRequest) returns (ResolveResponse) {}
+    // DAGService
 
-    rpc Tree(TreeRequest) returns (TreeResponse) {}
+    rpc AddNode(AddNodeRequest) returns (AddNodeResponse) {}
 
-    // BlockStore() returns a blockstore.Blockstore that provides the following
```
Comment (asutula) [inline ipfs-lite/ipfs_lite.proto:201]: Removed all the extra BlockStore methods because they are mostly redundant to the Node methods and not the primary way to use ipfs-lite

File: server/server.go [hunk 20]
```diff
@@ -0,0 +1,393 @@
...
+}
+
+func (s *ipfsLiteServer) AddNode(ctx context.Context, req *pb.AddNodeRequest) (*pb.AddNodeResponse, error) {
+	cid, err := cid.Decode(req.Block.GetCid())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: there is also just blocks.NewBlock()
+	block, err := blocks.NewBlockWithCid(req.Block.GetRawData(), cid)
```
Comment (asutula) [inline server/server.go]: At this point, I have the client sending a cid string and the raw node bytes. I'm using both of those bits of info to reconstruct the Block and Node here on the server. I'm a bit confused because you can also do this without the cid (I'm not sure how it knows how the bytes are encoded in that case). Anyway, open to suggestions about the right way to do this.

Comment (asutula) [reply server/server.go]: The raw bytes from the client can be encoded using cbor or Protobufs or any other supported format I guess.

Comment (asutula) [reply server/server.go]: This also raises questions about how clients in different languages will create the raw bytes to send. There looks to be a ipld cbor lib for Java but not Obj C, for example.

File: server/server.go [hunk 21]
```diff
@@ -0,0 +1,393 @@
...
+	return &pb.AddNodesResponse{}, nil
+}
+
+func (s *ipfsLiteServer) GetNode(ctx context.Context, req *pb.GetNodeRequest) (*pb.GetNodeResponse, error) {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+	// TODO: use session() NodeGetter or Peer NodeGetter methods directly?
+	node, err := s.peer.Get(ctx, cid)
```
Comment (asutula) [inline server/server.go:200]: I'm calling the peer's get node method directly, but there is also a `Session()` method that returns a "session-based NodeGetter". Not sure what the difference is or which I should use.

File: server/server.go [hunk 22]
```diff
@@ -0,0 +1,393 @@
...
+		return nil, err
+	}
+
+	node, err := s.peer.Get(ctx, cid)
+	if err != nil {
+		return nil, err
+	}
+
+	// ToDo: Figure out how to use this _ interface{} data
+	_, remainingPath, err := node.Resolve(req.GetPath())
```
Comment (asutula) [inline server/server.go]: Not sure how to deal with the data returned by resolve. It can be anything from a `Link` to a number or string or whatever data happens to be resolved. How do we represent that data and get it to the client?

**Comment by asutula** [comment]
Ok this is ready for a full review now.

File: server/server.go (hunk 20 above)
Comment (sanderpick) [reply server/server.go]: I believe the cid comes from the raw bytes, which use multicodec to inform readers how they are encoded: https://github.com/multiformats/multicodec

Comment (sanderpick) [reply server/server.go]: I'm always forgetting which multi thing does what. This helps: https://github.com/multiformats/multiformats/blob/master/img/multiformats.001.jpg

File: server/server.go (hunk 21 above)
Comment (sanderpick) [reply server/server.go:200]: Cool, good enough for now. I've never used the sessions but they could be handy at some point. The idea is to reduce bandwidth usage if you need to get a bunch of dags (I think): 

From the `go-blockservice`
```
// NewSession creates a new session that allows for
// controlled exchange of wantlists to decrease the bandwidth overhead.
// If the current exchange is a SessionExchange, a new exchange
// session will be created. Otherwise, the current exchange will be used
// directly.
```

File: server/server.go (hunk 22 above)
Comment (sanderpick) [reply server/server.go]: What about providing a tighter API like `ResolveLink`, `ResolveNumber`, `ResolveString`, `ResolveBytes`, etc.?

Comment (sanderpick) [reply server/server.go]: (you'd error if the type check failed)

File: server/server.go [hunk 23]
```diff
@@ -0,0 +1,393 @@
...
+		return err
+	}
+
+	buffer := make([]byte, getFileChunkSize)
+	for {
+		size, err := reader.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		srv.Send(&pb.GetFileResponse{Chunk: buffer[:size]})
```
Comment (sanderpick) [inline server/server.go:132]: Cool!

File: server/server.go (hunk 20 above)
Comment (sanderpick) [reply server/server.go]: Looking at this again, it's probably equally common to add a block without the cid. In that case, it's just generated automatically. (without CID -> you added some new data, with cid -> you're adding a block you received by requesting it from the network using its cid, as in, you already have it)

File: server/server.go (hunk 22 above)
Comment (sanderpick) [reply server/server.go]: Oh, I see there's already a `ResolveLink`

**Review by sanderpick — APPROVED** [review]
Hell yeah.

File: server/server.go (hunk 22 above)
Comment (asutula) [reply server/server.go]: I'm not sure if it's common to resolve anything but a Link? Is it?

Comment (asutula) [reply server/server.go]: In theory, you normally just fetch the Node you're interested in and unmarshall its data into some known model data type

File: server/server.go (hunk 20 above)
Comment (asutula) [reply server/server.go]: So, should we just treat the cid param as optional and call the corresponding method depending on if it is set?

File: server/server.go (hunk 21 above)
Comment (asutula) [reply server/server.go:200]: Sounds good.

File: server/server.go (hunk 22 above)
Comment (asutula) [reply server/server.go]: What do you think about my theory @sanderpick? If it's the case that you'd normally not want/need to resolve anything but `Link`s, we could just not expose this `Resolve` method (in fact, it's not even exposed in the top level IPFS-lite api, I just chose to dig into and expose the `Node`'s api here).

File: server/server.go (hunk 20 above)
Comment (sanderpick) [reply server/server.go]: SGTM

File: server/server.go (hunk 22 above)
Comment (sanderpick) [reply server/server.go]: Yep, makes sense - come to think of it, I've never used Resolve at all... so, +1 for leaving it out.

Comment (asutula) [reply server/server.go]: Sweeeeet

File: server/server.go [hunk 24]
```diff
@@ -0,0 +1,377 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	paths := node.Tree(req.GetPath(), int(req.GetDepth()))
+
+	return &pb.TreeResponse{Paths: paths}, nil
+}
+
+func pbBlockToBlock(pbBlock *pb.Block) (blocks.Block, error) {
```
Comment (asutula) [inline server/server.go:319]: @sanderpick here's a method that's used in `AddNode` and `AddNodes` that checks if the request includes a cid and uses it if so. At this time, I have a test that doesn't send a cid. This method runs with no problem and returns a Block created only from the raw data (which is some cbor node data). See my next comment in `AddNode` about an error.

File: server/server.go [hunk 25]
```diff
@@ -0,0 +1,377 @@
...
+	return &pb.HasBlockResponse{HasBlock: hasBlock}, nil
+}
+
+func (s *ipfsLiteServer) AddNode(ctx context.Context, req *pb.AddNodeRequest) (*pb.AddNodeResponse, error) {
+	block, err := pbBlockToBlock(req.GetBlock())
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := format.Decode(block)
```
Comment (asutula) [inline server/server.go:157]: So I get my block back above (created without specifying a cid) and this call to `Decode` fails with the error: ```failed to add node: rpc error: code = Unknown desc = failed to decode Protocol Buffers: incorrectly formatted merkledag node: unmarshal failed. unexpected EOF```

If I change the test to specify the cid, this all works no problem.

File: server/server.go (hunk 24 above)
Comment (asutula) [reply server/server.go:319]: For context, here is the test code that creates the Node and calls the AddNode method:
```
func createNode(data map[string]interface{}) (*cbor.Node, error) {
	codec := uint64(multihash.SHA2_256)
	return cbor.WrapObject(data, codec, multihash.DefaultLengths[codec])
}

func TestAddNode(t *testing.T) {
	...
	node0data := map[string]interface{}{
		"name": "node0",
	}
	node0, err := createNode(node0data)
	if err != nil {
		t.Fatalf("failed to create node: %v", err)
	}

	block0 := pb.Block{
		// Cid:     node0.Cid().String(),
		RawData: node0.RawData(),
	}

	_, err = client.AddNode(ctx, &pb.AddNodeRequest{Block: &block0})
	if err != nil {
		t.Fatalf("failed to add node: %v", err)
	}
	...
}
```
Notice the Cid is commented out in the data that's sent to AddNode

File: server/server.go (hunk 25 above)
Comment (asutula) [reply server/server.go:157]: It's strange that it thinks the data is Protobuf data when I encoded it with cbor.
Screen Shot 2019-11-01 at 5 23 49 PM
That screen shot is me stepping into the decoding process. You can see that the detected decoder is for Protobuf data.

File: server/server.go (hunk 24 above)
Comment (sanderpick) [reply server/server.go:319]: Hmm, I could be mistaken about the block creation if it's not working. I guess RawData is serialized cbor, meaning there's nothing about it the self-declares it _is_ cbor, like a multicodec.

Comment (sanderpick) [reply server/server.go:319]: What is the IPFS Lite method you're exposing here?

Comment (asutula) [reply server/server.go:319]: This would be the `AddNode` and `AddNodes` methods

Comment (asutula) [reply server/server.go:319]: So @sanderpick, I guess the bytes themselves aren't self describing -- it's the cid that describes how the bytes are encoded. I guess if you don't provide a cid, the default assumption is the bytes are Protobuf?

Comment (sanderpick) [reply server/server.go:319]: Right, seems that way.

Comment (sanderpick) [reply server/server.go:319]: Can you point me to the AddNode method on the ipfs lite repo?

Comment (sanderpick) [reply server/server.go:319]: The typical way to add nodes if with a dag service, which the ipfs lite instance itself becomes. Then `Add` and `AddMany` are at your disposal. 

https://github.com/ipfs/go-ipld-format/blob/master/merkledag.go#L30

Comment (asutula) [reply server/server.go:319]: Yea those are the exact methods I'm exposing.

Comment (asutula) [reply server/server.go:319]: Ok, so if you `AddNode` or `AddNodes` without passing a Cid, the node data must be ProtoNode. If you pass a Cid, then you're free to use cbor or Protobuf or whatever supported format.
